src/networking.c

```diff
@@ -3238,17 +3307,158 @@ int canParseCommand(client *c) {
+                hashtableIncrementalFindState *incr_state = &incr_states[num++];
+                robj *keyobj = argv[result.keys[i].pos];
+                hashtableIncrementalFindInit(incr_state, ht, keyobj->ptr);
+            }
+        }
+    }
+    getKeysFreeResult(&result);
+    return num;
+}
+
+/* Prefetches the keys for the commands queued up in the client. */
+static void prefetchCommandQueueKeys(client *c) {
+    if (c->read_flags & READ_FLAGS_PREFETCHED) return;
+
+    /* Prefetching states */
+    const int max_keys = 32;
+    const int soft_max_keys = 16; /* Threshold to continue to the next command
+                                   * and prefetch its keys */
+    int num_keys = 0;
+    hashtableIncrementalFindState key_incr_states[max_keys];
+
+    /* Lookup commands and add keys to incremental find batch. */
+    if (c->read_flags & READ_FLAGS_PARSING_COMPLETED && c->argc != 0) {
+        c->read_flags |= READ_FLAGS_PREFETCHED;
+        if (!c->io_parsed_cmd) {
+            struct serverCommand *cmd = lookupCommand(c->argv, c->argc);
+            if (!cmd || !commandCheckArity(cmd, c->argc, NULL)) {
+                /* Wrong arity. This case is handled later. */
+            } else {
+                c->io_parsed_cmd = cmd;
+            }
+        }
+        if (c->io_parsed_cmd) {
+            num_keys = addKeysToIncrFindBatch(c, c->io_parsed_cmd, c->argv, c->argc,
+                                              key_incr_states, num_keys, max_keys);
+        }
+    }
+
+    cmdQueue *queue = &c->cmd_queue;
+    for (int i = queue->off; i < queue->len; i++) {
+        commandParserState *st = &queue->cmds[i];
+        if (!(st->read_flags & READ_FLAGS_PARSING_COMPLETED) || c->argc == 0) continue;
+        if (num_keys >= soft_max_keys) break;
+        st->read_flags |= READ_FLAGS_PREFETCHED;
+        if (!st->cmd) {
+            struct serverCommand *cmd = lookupCommand(st->argv, st->argc);
+            /* Wrong arity. Reset command so it will be handled later. */
+            if (!cmd || !commandCheckArity(cmd, st->argc, NULL)) continue;
+            st->cmd = cmd;
+        }
+        num_keys = addKeysToIncrFindBatch(c, st->cmd, st->argv, st->argc,
+                                          key_incr_states, num_keys, max_keys);
+    }
+    if (num_keys <= 1) return; /* No point to batch-lookup a single key */
+
+    /* Batch-lookup the keys. */
+    int not_complete_count;
+    do {
+        not_complete_count = 0;
+        for (int i = 0; i < num_keys; i++) {
+            not_complete_count += hashtableIncrementalFindStep(&key_incr_states[i]);
+        }
+    } while (not_complete_count != 0);
+
+    /* Prefetch value pointers. */
+    for (int i = 0; i < num_keys; i++) {
+        void *entry;
+        if (hashtableIncrementalFindGetResult(&key_incr_states[i], &entry)) {
+            robj *val = entry;
+            /* TODO? Prefetch all types and encodings except OBJ_ENCODING_EMBSTR
+             * and OBJ_ENCODING_INT. */
```

Comment:
This seems right? Everything else should be prefetchable for values.

Reply:
Yes we could. Only some commands like EXISTS don't need the value, and commands like UNLINK access it only in the background threads, but most others do. Maybe we could look at command flags or key specs flags to determine this... We could move this idea to a separate issue/PR if it's non-trivial.

I'd like to unify this with prefetching code with the one in memory_prefetch.c and refactor that. I will do it.